runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerEnvironmentFactory.java

```diff
@@ -0,0 +1,136 @@
+      Supplier<String> idGenerator) {
+    return new DockerEnvironmentFactory(
+        docker,
+        controlServiceServer,
+        loggingServiceServer,
+        retrievalServiceServer,
+        provisioningServiceServer,
+        idGenerator,
+        clientSource);
+  }
+
+  private final DockerWrapper docker;
+  private final GrpcFnServer<FnApiControlClientPoolService> controlServiceServer;
+  private final GrpcFnServer<GrpcLoggingService> loggingServiceServer;
+  private final GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer;
+  private final GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer;
+  private final Supplier<String> idGenerator;
+  private final ControlClientSource clientSource;
+
+  private DockerEnvironmentFactory(
+      DockerWrapper docker,
+      GrpcFnServer<FnApiControlClientPoolService> controlServiceServer,
+      GrpcFnServer<GrpcLoggingService> loggingServiceServer,
+      GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer,
+      GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer,
+      Supplier<String> idGenerator,
+      ControlClientSource clientSource) {
+    this.docker = docker;
+    this.controlServiceServer = controlServiceServer;
+    this.loggingServiceServer = loggingServiceServer;
+    this.retrievalServiceServer = retrievalServiceServer;
+    this.provisioningServiceServer = provisioningServiceServer;
+    this.idGenerator = idGenerator;
+    this.clientSource = clientSource;
+  }
+
+  /** Creates a new, active {@link RemoteEnvironment} backed by a local Docker container. */
+  @Override
+  public RemoteEnvironment createEnvironment(Environment environment) throws Exception {
+    String workerId = idGenerator.get();
+
+    // Prepare docker invocation.
+    Path workerPersistentDirectory = Files.createTempDirectory("worker_persistent_directory");
+    Path semiPersistentDirectory = Files.createTempDirectory("semi_persistent_dir");
+    String containerImage = environment.getUrl();
+    // TODO: https://issues.apache.org/jira/browse/BEAM-4148 The default service address will not
+    // work for Docker for Mac.
+    String loggingEndpoint = loggingServiceServer.getApiServiceDescriptor().getUrl();
+    String artifactEndpoint = retrievalServiceServer.getApiServiceDescriptor().getUrl();
+    String provisionEndpoint = provisioningServiceServer.getApiServiceDescriptor().getUrl();
+    String controlEndpoint = controlServiceServer.getApiServiceDescriptor().getUrl();
+    List<String> args =
+        Arrays.asList(
+            "-v",
+            // TODO: Mac only allows temporary mounts under /tmp by default (as of 17.12).
+            String.format("%s:%s", workerPersistentDirectory, semiPersistentDirectory),
+            // NOTE: Host networking does not work on Mac, but the command line flag is accepted.
+            "--network=host",
+            containerImage,
+            String.format("--id=%s", workerId),
```

Comment:
<!--thread_id:cc_182928850_t; commit:fb7d2bd7ba1bb27186e8d2dda861acd08666a0c1; resolved:1-->
<!--section:context-quote-->
> **angoenka** wrote:
> How will we keep it in sync with the boot.go file?

<!--section:body-->
No need to. The container contract means that the invocation will always be the same, even between different sdks. I linked https://issues.apache.org/jira/browse/BEAM-4148 above specifically to address Mac-specific changes that need to happen. However, note that all of these changes are runner-side. It is ultimately the runner that is responsible for passing a usable endpoint to SDK harnesses.